src/Form/Form.js

```diff
@@ -221,16 +221,25 @@ class Form {
         const selector = this.matching.cssSelector('SUBMIT_BUTTON_SELECTOR')
         const allButtons = /** @type {HTMLElement[]} */([...this.form.querySelectorAll(selector)])
 
-        const likelySubmitButton = allButtons.find(isLikelyASubmitButton)
-        if (likelySubmitButton) return [likelySubmitButton]
-
-        return allButtons.filter((button) => {
-            const content = button.textContent || ''
-            const ariaLabel = button.getAttribute('aria-label') || ''
-            const title = button.title || ''
-            // trying to exclude the little buttons to show and hide passwords
-            return !/password|show|toggle|reveal|hide/i.test(content + ariaLabel + title)
-        })
+        return allButtons
+            .filter(isLikelyASubmitButton)
+            .filter((button) => {
+                const content = button.textContent || ''
+                const ariaLabel = button.getAttribute('aria-label') || ''
+                const title = button.title || ''
+                // exclude buttons to show/hide passwords, cancel buttons, social login buttons
+                return !/facebook|twitter|google|apple|cancel|password|show|toggle|reveal|hide/i.test(content + ariaLabel + title)
+            })
+            // filter out buttons of the wrong type - login buttons on a signup form, signup buttons on a login form
+            .filter((button) => {
+                if (this.isLogin) {
+                    return !/sign.?up/i.test(button.textContent || '')
+                } else if (this.isSignup) {
+                    return !/(log|sign).?([io])n/i.test(button.textContent || '')
+                } else {
+                    return true
+                }
```

Comment:
Mhm, so this is a little contrived. If you invert the regexes (login for signup and viceversa) you don't need the negations. We may also want to use the same strategy for button content we're using above (with `value`, `title`, `ariaLabel`). Is it worth extracting a function for something like `getButtonText` in autofill-utils.js or somewhere else?

Reply:
I don't think this is true, because the regexes aren't inverses - there are buttons that match neither. Currently, on a login form, all buttons are allowed unless they match "sign up". If you do this:
```
if (this.isLogin) {
                    return /(log|sign).?([io])n/i.test(button.textContent || '')
                } else if (this.isSignup) {
                    return /sign.?up/i.test(button.textContent || '')
                } else {
                    return true
                }
```
Then only buttons matching "log in" are allowed, which causes a bunch of test failures.

(On an actual login form, it would be reasonable to assume that the submit button contains "log in". But formAnalyzer classifies everything as either login/sign up, so there are all kinds of other forms being classified as a login form).

Reply:
You're right, of course 🤦‍♂️.